### ISSUE TITLE
Support keep stride for neg with requires_grad=False

### ISSUE THREAD

**Issue by boeddeker** [pytorch/pytorch#44182]
## 🚀 Feature

Support that the output stride is identical to the input stride for `__neg__`, at least, when `requires_grad=False`.
Advantage: Less memory consumption and fewer computations.

The code
```python
import torch
print(torch.__version__)
x = torch.tensor([[1., 2, 3], [4., 5, 6]], device='cuda:0')
c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
print(c.stride())
print((-c).stride())
```
produces
```
1.7.0.dev20200807
(3, 1, 1)
(4, 2, 1)
```
while I want
```
1.7.0.dev20200807
(3, 1, 1)
(3, 1, 1)
```

## Motivation

I have a small tensor, but for the mathematical operation I have to use a large view of the tensor.
The mathematical operations, that I use, require a "conjugate" operation.
I want to realize this with taking the negative of the imaginary part (At the moment I use https://github.com/kamo-naoyuki/pytorch_complex).
But this is at the moment to expensive, because the real size of the tensor is a few GB.

My real motivation is to make https://github.com/fgnt/nara_wpe/pull/47 more memory efficient.
But that example is more complicated.

## Pitch

Keep the input stride when calling: `- tensor`

## Alternatives

Calculating the view two times.

**Comment by ngimel** [comment]
While for unary functions you technically can preserve the input strides, the general contract in pytorch has been that operations produce non-self-overlapping outputs, and, in fact, pytorch errors out when trying to operate in-place on a self-overlapping tensors (*in some cases, in general this is a hard problem). This behavior is also consistent with numpy. Power users who know what they are doing can use the workaround of computing 2 views.

**Comment by boeddeker** [comment]
@ngimel I know that it is non-trivial in general. But for some operations like element wise unary functions it is.
And numpy has the property of keeping the strides for some operations, like element wise unary functions.

> Power users who know what they are doing can use the workaround of computing 2 views.

Can you explain me, how that works?
I am not able to do that. Here is the example that I tried:

```python
    >>> import torch
    >>> torch.__version__
    '1.8.0.dev20201022+cpu'
    >>> x = torch.tensor([[1., 2, 3], [4., 5, 6]])
    >>> c = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
    >>> print(c)
    tensor([[[1., 2.],
             [2., 3.]],
    <BLANKLINE>
            [[4., 5.],
             [5., 6.]]])
    >>> print(-c)
    tensor([[[-1., -2.],
             [-2., -3.]],
    <BLANKLINE>
            [[-4., -5.],
             [-5., -6.]]])
    >>> y = torch.tensor([[1., 2, 3], [4., 5, 6]])
    >>> d = torch.as_strided(x, size=[2, 2, 2], stride=[3, 1, 1])
    >>> torch.neg(c, out=d)
    tensor([[[-1.,  2.],
             [ 2., -3.]],
    <BLANKLINE>
            [[-4.,  5.],
             [ 5., -6.]]])
    >>> c
    tensor([[[-1.,  2.],
             [ 2., -3.]],
    <BLANKLINE>
            [[-4.,  5.],
             [ 5., -6.]]])
    >>> d
    tensor([[[-1.,  2.],
             [ 2., -3.]],
    <BLANKLINE>
            [[-4.,  5.],
             [ 5., -6.]]])
```
Note: `d` is the negative value of `c`, but `c` got modified (I thought `torch.neg` is not an inplace operation, if it should, this is a bug).

**Comment by ngimel** [comment]
`c` got modified because both `c` and `d` point to the same storage (`x`), so when you are doing `torch.neg(c, out=d)` it's effectively inplace operation. If your tensor occupies all underlying storage, you'd do something like
```
tmp = torch.empty((), device=c.storage().device, dtype=c.storage().dtype)
tmp.set_(c.storage(), 0, (c.storage().size(),))
tmp = -tmp
d = torch.as_strided(tmp, size=[2,2,2], stride=[3,1,1])
```
If it's some view of the underlying storage, you'd need to properly set storage offset (currently hardcoded to 0) and extent occupied by your tensor (currently c.storage().size)). It won't work unfortunately if your strided tensor occupies discontiguous memory region.

**Comment by boeddeker** [comment]
Thank you. I used too much copy and paste.
Your example is very helpful and increased my knowledge about pytorch. 
Based on your example, I found, that the following works:
```python
def neg(tensor):
    return torch.as_strided(
        -torch.Tensor(c.storage()),
        size=tensor.size(),
        stride=tensor.stride(),
        storage_offset=c.storage_offset())
```

**Comment by ngimel** [comment]
It will work if tensor uses the whole storage, it won't for something like
```
a=torch.randn(64,32)
b,c=a.chunk(2) 
#c1 = some_fancy_as_strided(c)
#how to do -c1 now? c.storage() will give a full storage that's also backing b
```

**Comment by boeddeker** [comment]
Yes, I agree, that it does not work in all cases.
In my case the storage is amplitudes smaller than the virtual tensor size.

> how to do -c1 now? c.storage() will give a full storage that's also backing b

In your example, a force contiguous is the better choice.
But when you do some broadcasting on `c`, it may be better to copy the storage that also contains `b`

A simple heuristic would be:
```python
if tensor.nelement() > tensor.storage().size():
    # keep original memory layout
else:
    # force contiguous
```
but I don't know if such a comparison would be acceptable in some low level code
and if it is too much magic for users.

For my problem the workaround with the storage is ok.
